Update keepalived script and refactor related part in labkeeper

bzhaoopenstack opened the pull request:
For this loop.
The following tests are passed.
1. unecessary serivice down, post an issue, if the issue is timeout, do switch.
2. necessary service down, post an issue, do switch.
3. master host down/ keepalived is down(heartbeat fail to update, then timeout), post an issue, do switch.

Commit:
Update keepalived script and labkeeper code

This patch add a totally new keepalived script and this script will run
as a healthcheck for fit the most cases in HA scenario. Also both
ha-master and ha-slave will use the same script to check local services
and raise the M/S switch.

Commit:
rm bad dir

Commit:
rm bad code

liusheng commented on playbooks/keepalived.yaml:
```diff
@@ -45,62 +35,34 @@
 
     - name: install script dependency
       pip:
-        name: ['PyGithub', 'argparse', 'requests', 'html', 'datetime']
```
please use a list of packages for `pip`, the with_items of pip installation usage has been deprecated.

bzhaoopenstack replied:
done

wangxiyuan commented on etc/keepalived/kp_local_loop.py:
```diff
@@ -0,0 +1,581 @@
+DOMAIN_NAME = 'openlabtesting.org'
+DOMAIN_NAME1 = 'status.openlabtesting.org'
+DOMAIN_NAME2 = 'logs.openlabtesting.org'
+DOMAIN_NAME_BK = 'logs-bak.openlabtesting.org'
```
what's this used for?

bzhaoopenstack replied:
Used for change DNS

wangxiyuan commented on etc/keepalived/kp_local_loop.py:
```diff
@@ -0,0 +1,581 @@
+    if not GLOBAL_ZK:
+        cfg = configparser.ConfigParser().read(ZK_CLI_CONF)
+        zk_cli = zk.ZooKeeper(cfg)
+        zk_cli.connect()
```
you must disconnect zk as well after every time you connect to zk.

bzhaoopenstack replied:
Why? I prefer to use a global single session during access z, like we use zkCli.sh..I test it, it's OK to do this..  Or is there any other issues I need to fix if don't close the connection?

wangxiyuan replied:
If you disconnect zk by hand in the end of the script, zk will log info that `session is disconnected`.

If you don't, when the script is done, zk will log warning that `session is lost` which is not a correct way.

bzhaoopenstack replied:
Ok, so you are agree with the single session, but need close the session at the end of script. Right?

wangxiyuan replied:
++

wangxiyuan commented on etc/keepalived/kp_local_loop.py:
```diff
@@ -0,0 +1,581 @@
+
+    global GLOBAL_NODE
+    GLOBAL_NODE = local_node
+    return GLOBAL_NODE
```
no need to set GLOBAL_NODE here, just return `local_node`

bzhaoopenstack replied:
;-)

wangxiyuan replied:
just return local_node here and then set `GLOBAL_NODE=load_from_zk()` at L275, so that you done't need declare `GLOBAL_NODE` once again here.

bzhaoopenstack replied:
done

wangxiyuan commented on etc/keepalived/kp_local_loop.py:
```diff
@@ -0,0 +1,581 @@
+                      zk_node['heartbeat'], zk_node['alarmed'],
+                      zk_node['status'])
+
+    for service in zk_cli.list_services(node=node_name):
```
`list_services(node_name_filter=node_name)`

bzhaoopenstack replied:
Thanks, I realize the interface had been updated.

wangxiyuan commented on etc/keepalived/kp_local_loop.py:
```diff
@@ -0,0 +1,581 @@
+    def is_alarmed_timeout(self):
+        over_time = parse_isotime(
+            self.alarmed_at) + datetime.timedelta(
+            days=2)
```
it's good to be configurable for the `days`, hour is better. We can save it in zk

bzhaoopenstack replied:
Right, We can support it by storing necessary info into zk.

Commit:
Update local check script

Fix some issue according to the real test env.

wangxiyuan commented on etc/keepalived/kp_local_loop.py:
```diff
@@ -0,0 +1,665 @@
+def load_from_zk():
+    zk_cli = get_zk_cli()
+    node_name = get_local_hostname()
+    zk_cli.connect()
```
you have connected to zk at L248, why connect it again?

bzhaoopenstack replied:
Correct, remove it

wangxiyuan commented on etc/keepalived/kp_local_loop.py:
```diff
@@ -0,0 +1,665 @@
+
+def load_from_zk():
+    zk_cli = get_zk_cli()
+    node_name = get_local_hostname()
```
I don't think this is worth to use a  new function.

bzhaoopenstack replied:
Now, the node_name designation is different with openlabcmd format node name. So I just split this out for custmized the hostname to fit the gap.

bzhaoopenstack replied:
I add a TODO on this function

wangxiyuan commented on etc/keepalived/kp_local_loop.py:
```diff
@@ -0,0 +1,665 @@
+
+
+def treat_single_service(service_obj, node_obj):
+    if node_obj.is_maintained():
```
this can  be removed now. so as `is_maintained()`

bzhaoopenstack replied:
done

wangxiyuan commented on etc/keepalived/kp_local_loop.py:
```diff
@@ -0,0 +1,665 @@
+            zk_cli = get_zk_cli()
+            updated_svc = zk_cli.update_service(
+                self.name, self.node_role, self.node_type,
+                restarted=True, status='restarting')
```
I didn't find the code to set 'restarting' to 'up' or 'down'?

bzhaoopenstack replied:
Please check in L145， each loop， if the current restart is different with restarting, it will store the new status in zk.

wangxiyuan replied:
Emm, right. So you **update** the service in `is_abnormal` function which I assume is a function for **check**. It's a little strange.

wangxiyuan commented on etc/keepalived/kp_local_loop.py:
```diff
@@ -0,0 +1,665 @@
+
+    if service_obj.is_abnormal():
+        if service_obj.restarted:
+            if node_obj.is_necessary_service(service_obj.name):
```
service_obj has a property called `is_necessary`, why don't check it directly? I have to say that the list `necessary_service_names` in Service class is redundant

bzhaoopenstack replied:
Yeah, this code need to be refactor, this code is fit for the original openlabcmd code.

bzhaoopenstack replied:
done

Commit:
Address Yuan God's comments

This patch fix Yuan's review. Also fix the oppo check during master down cases

wangxiyuan commented on etc/keepalived/kp_local_loop.py:
```diff
@@ -0,0 +1,693 @@
+
+    # At the end of the script, we'd better to close zk session.
+    zk_cli = get_zk_cli()
+    zk_cli.close()
```
`zk_cli.disconnect()`

bzhaoopenstack replied:
done ;-)

wangxiyuan commented on etc/keepalived/kp_local_loop.py:
```diff
@@ -0,0 +1,693 @@
+        current_time = datetime.datetime.utcnow().replace(tzinfo=iso8601.UTC)
+        return current_time > over_time
+
+    def is_necessary_service(self, service_name):
```
this function can be removed

bzhaoopenstack replied:
done

wangxiyuan commented on etc/keepalived/kp_local_loop.py:
```diff
@@ -0,0 +1,693 @@
+def setup_necessary_services_and_check(node_obj):
+    for service in node_obj.necessary_service_names:
+        run_systemctl_command(SYSTEMCTL_START, service)
+
```
```
for service in node_obj.services:
    if service['is_necessary']:
        run_systemctl_command(SYSTEMCTL_START, service)
```

Then the `necessary_service_names` property can be removed totally.

bzhaoopenstack replied:
done

Commit:
Address Yuan God's comments

1. use zk_cli.disconnect
2. remove necessary_services_name in Node obj. And refactor the related part.

wangxiyuan reviewed (approved):
LGTM, Let's merge it and take a fully test.

Commit:
run kp with test account

Commit:
update kp loop for change dns logic for fit test account

Commit:
update pip packages

Commit:
fix kp restart

Commit:
fix kp

Commit:
remove oppo_ip in keepalived conf

Commit:
Fix kp for update status after fix issue towards services

1. recover the service status if we try to restart
2. recover the service status if fix manually.
